connector/loganomalyconnector/anomaly.go

```diff
@@ -0,0 +1,285 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package loganomalyconnector provides an OpenTelemetry collector connector that detects
+// anomalies in log throughput using statistical analysis. It monitors the rate of incoming
+// logs and generates alerts when significant deviations from the baseline are detected,
+// using both Z-score and Median Absolute Deviation (MAD) methods for anomaly detection.
+package loganomalyconnector
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+// Sample represents a single measurement of log throughput at a specific point in time.
+// It pairs a timestamp with the corresponding rate of logs per minute for that sample period.
+type Sample struct {
+	timestamp time.Time
+	rate      float64
+}
+
+// Statistics holds statistical measures calculated from a set of log rate samples.
+// These statistics are used to establish a baseline and detect anomalies in log throughput.
+type Statistics struct {
+	mean    float64
+	stdDev  float64
+	median  float64
+	mad     float64
+	samples []float64
+}
+
+// AnomalyStat contains information about a detected log throughput anomaly,
+// including the type of anomaly, baseline statistics, and deviation metrics.
+type AnomalyStat struct {
+	anomalyType    string
+	baselineStats  Statistics
+	currentRate    float64
+	zScore         float64
+	madScore       float64
+	percentageDiff float64
+	timestamp      time.Time
+}
+
+// takeSample calculates and stores a new rate sample
+func (d *Detector) takeSample(now time.Time) {
+	duration := now.Sub(d.currentBucket.start).Minutes()
+	if duration < (1.0 / 60.0) {
+		return
+	}
+	rate := float64(d.currentBucket.count) / duration
+
+	d.rateHistory = append(d.rateHistory, Sample{
+		timestamp: now,
+		rate:      rate,
+	})
+
+	d.currentBucket.count = 0
+	d.currentBucket.start = now
+	d.lastSampleTime = now
+
+	d.pruneLogs()
+	if anomaly := d.checkForAnomaly(); anomaly != nil {
+		d.logAnomaly(anomaly)
+	}
+}
+
+// pruneLogs performs cleanup on log count buffer
+func (d *Detector) pruneLogs() {
+	if len(d.rateHistory) == 0 {
+		return
+	}
+
+	cutoffTime := time.Now().Add(-d.config.MaxWindowAge)
+	idx := sort.Search(len(d.rateHistory), func(i int) bool {
+		return d.rateHistory[i].timestamp.After(cutoffTime)
+	})
+	if idx > 0 {
+		d.rateHistory = d.rateHistory[idx:]
+	}
+
+	// in the case we have more logs than specified for our buffer
+	if len(d.rateHistory) > d.config.EmergencyMaxSize {
```

Comment:
If we're maintaining a maximum window age and buckets of fixed durations, how would this condition come about?